packages/editor/src/lib/codemirror.ts

```diff
@@ -0,0 +1,173 @@
+  errorMessageID?: string | undefined;
+  isInvalid?: boolean | undefined;
+  readonly?: boolean;
+  onFocus?: (event: FocusEvent) => void;
+  onBlur?: (event: FocusEvent) => void;
+  diffOriginal?: string | undefined;
+}
+
+/** Create an editor StateField and updater function. */
+const createSource = <T>(): [
+  source: StateField<T | undefined>,
+  update: (view: EditorView | undefined, nextValue: T | undefined) => void,
+] => {
+  const updateEffect = StateEffect.define<T | undefined>();
+
+  const source = StateField.define<T | undefined>({
+    create: () => undefined,
+    update: (prevValue, transaction) => {
+      let nextValue = prevValue;
+
+      for (const effect of transaction.effects) {
+        if (effect.is(updateEffect)) {
+          nextValue = effect.value;
+        }
+      }
+
+      return nextValue;
+    },
+  });
+
+  const update = (view: EditorView | undefined, nextValue: T | undefined) => {
+    view?.dispatch({ effects: [updateEffect.of(nextValue)] });
+  };
+
+  return [source, update];
+};
+
+const [readonly, setReadOnly] = createSource<boolean>();
+const [label, setLabel] = createSource<string>();
+const [errorMessageID, setErrorMessageID] = createSource<string>();
+const [isInvalid, setIsInvalid] = createSource<boolean>();
+
+const setDiffOriginal = (
+  view: EditorView | undefined,
+  nextOriginal: string | undefined
+) => {
+  if (nextOriginal !== undefined) {
+    view?.dispatch({
+      effects: updateOriginalDoc.of({
+        doc: view.state.toText(nextOriginal),
+        changes: ChangeSet.empty(0),
+      }),
+    });
+  }
+};
+
+/**
+ * Create a CodeMirror extension list for a JSON editor.
+ *
+ * Adds various state fields and effects so that props changes
+ * can be hooked to the editor without needing to completely
+ * reconfigure or recreate the extensions list.
+ *
+ * @param options Initial values for editor state
+ * @returns The extension list and hooks to reactively update the state
+ *
+ */
+
+export const createJSONExtensions = (
+  options: ExtensionOptions
+): readonly Extension[] => {
+  const extensions = [
+    readonly.init(() => options.readonly),
+    label.init(() => options.label),
+    errorMessageID.init(() => options.errorMessageID),
+    isInvalid.init(() => options.isInvalid),
+    basicSetup,
+    json(),
+    EditorView.contentAttributes.compute(
+      [label, errorMessageID, isInvalid],
+      (state) => {
+        const computedAttributes: Record<string, string> = {};
+        const labelValue = state.field(label);
+        if (labelValue !== undefined) {
+          computedAttributes['aria-label'] = labelValue;
+        }
+        if (state.field(isInvalid) === true) {
+          computedAttributes['aria-invalid'] = 'true';
+        }
+        const errorMessageIDValue = state.field(errorMessageID);
+        if (errorMessageIDValue !== undefined) {
+          computedAttributes['aria-errormessage'] = errorMessageIDValue;
+        }
+        return computedAttributes;
```

Comment:
This used `stripUndefined` when it was in app. However, it does not seem worth it to bring over that func.

Reply:
Seems like your implementation here covers it, so SGTM.